### PR TITLE
[feature](json-function) add json_depth function

### LINKED ISSUE #21268: [Enhancement] Improve Json function for doris
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Now We have some _JSONB functions_ and  _JSON functions_ are not implement , specific semantics for each function below here can reference mysql

**NOTE** : _JSONB Functions_ , which has same function name with _JSON functions_ but with different implement which can be find in `be/src/vec/functions/function_jsonb.cpp ` . eg. json_parse alias jsonb_parse

### about json_path define:
https://modern-sql.com/blog/2017-06/whats-new-in-sql-2016#json-path

#### we have some JSONB functions which are not implement but JSON has already implement. below here
- [ ]  json_contains  @weijiew 
- [ ]  json_contains_path @weijiew 
- [ ]  json_quote json_unquote 
- [ ]  json_valid @hammerZh-Z 
- [ ]  json_array @yifei02 
- [ ]  json_object

#### below here we have both JSONB and JSON functions not implement .

- [ ]  ->>  ([->>](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#operator_json-inline-path))
- [ ] JSON_DEPTH ([JSON_DEPTH()](https://dev.mysql.com/doc/refman/8.0/en/json-attribute-functions.html#function_json-depth)) @niebayes 
- [ ] JSON_LENGTH ([JSON_LENGTH()](https://dev.mysql.com/doc/refman/8.0/en/json-attribute-functions.html#function_json-length)) @Xuxiaotuan 
- [ ] JSON_KEYS ([JSON_KEYS()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-keys)) @caicancai 
- [ ] JSON_SEARCH ([JSON_SEARCH()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-search)) 
- [ ] MERGE_OF ([MEMBER OF()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#operator_member-of))

- [ ] JSON_INSERT ([JSON_INSERT()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-insert)).  @JasonWen1 
- [ ] JSON_REPLACE ([JSON_REPLACE()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-replace)) @cjj2010 
- [ ] JSON_SET ([JSON_SET()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-set))

- [ ] JSON_OVERLAPS ([JSON_OVERLAPS()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-overlaps))
- [ ] JSON_MERGE_PATCH ([JSON_MERGE_PATCH()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-merge-patch))
- [ ] JSON_MERGE_PRESERVE ([JSON_MERGE_PRESERVE()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-merge-preserve)) @taptao 
- [ ] JSON_REMOVE ([JSON_REMOVE()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-remove))
- [ ] JSON_TABLE ([JSON_TABLE()](https://dev.mysql.com/doc/refman/8.0/en/json-table-functions.html#function_json-table)) @lovlyhao 
- [ ] JSON_STORAGE_SIZE ([JSON_STORAGE_SIZE()](https://dev.mysql.com/doc/refman/8.0/en/json-utility-functions.html#function_json-storage-size))
- [ ] JSON_STORAGE_FREE ([JSON_STORAGE_FREE()](https://dev.mysql.com/doc/refman/8.0/en/json-utility-functions.html#function_json-storage-free))

U can pick this according comment below:
"I pick "-> ->> " " and this task could assign to u for one week, if no pr or progress during one week I would re-assign others

### Solution

U can pick this according comment below:
"I pick "-> ->> " " and this task could assign to u for one week, if no pr or progress during one week I would re-assign others 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by niebayes**
## Proposed changes

Issue Number: close #21268 

## Further comments

Add `json_depth` function.

TODO: add benchmarking results for different JSON parser including simdjson parser and rapidjson parser.

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/function_json.cpp [hunk 1]
```diff
@@ -1128,6 +1129,95 @@ class FunctionJsonUnquote : public IFunction {
...
+        // Doris processes the input data in batch, aka. a block.
+        // a block consists of several cells where each cell is indexed by a row and a column numbers.
+        // for instance, the json string passed into this function is stored in a cell.
+        for (size_t i = 0; i < input_rows_count; ++i) {
+            if (nullable_input_col->is_null_at(i)) {
+                result_null_map->get_data()[i] = 1;
+                continue;
+            }
+
+            const ColumnString* col_str = check_and_get_column<ColumnString>(nullable_input_col);
```
Comment (amorynan) [inline be/src/vec/functions/function_json.cpp:1171]: make check_and_get_column<ColumnString>() out of forloop

File: be/src/vec/functions/function_json.cpp [hunk 2]
```diff
@@ -1128,6 +1129,95 @@ class FunctionJsonUnquote : public IFunction {
...
+            const std::string_view json_doc = col_str->get_data_at(i).to_string_view();
+
+            // the json elements are represented as a tree.
+            // `parse` takes in the json document and returns the root element of the tree if all goes well.
+            SimdJSONParser::Element root_element;
```
Comment (amorynan) [inline be/src/vec/functions/function_json.cpp:1176]: is simdjsonparser parse is better than others？ like rapidjson ？

**Comment by niebayes** [comment]
@amorynan I've updated the code by taking `check_and_get_column` out of the for loop.

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 55.26 seconds
 stream load tsv:          516 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          88.5 seconds inserted 10000000 Rows, about 112K ops/s
 storage size: 17164281709 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230706134736_clickbench_pr_173366.html

**Pull request closed without merging**